### ISSUE TITLE
Core 400S Pro not working

### ISSUE THREAD

**Issue by JakeScheck** [RaresAil/homebridge-levoit-air-purifier#2]
Thanks for the message on the other plug-ins page. This plugin looks very promising! Unfortunately it doesn’t seem to be finding my device. I have just a Levoit Core 400S set up in my VeSync account. I have installed the plugin and entered my credentials but in the status log the latest message I see is “Discovering Devices….” And I don’t see the air purifier show up in my Home app. Any ideas? Am I missing something in setup?

**Comment by RaresAil** [comment]
Do you have Discord so we can debug it? will setup a debug version and add logs to see what is happening there

**Comment by alockbox** [comment]
I’m having exactly the same experience with the Core 400s Pro “PlasmaPro 400s” (it’s branded two different ways on the box). This is essentially a 400s but with a positive/negative ionizer button as well. Nothing shows up in HomeKit. Most of the other plug-ins, my device does show even if all the features aren’t supported (one didn’t have auto mode, none show the ionizer, for ex.)

**Comment by RaresAil** [comment]
@alockbox Does your Air Purifier have auto mode? And i will prepare a branch with debugging logs and to can fix it

**Comment by alockbox** [comment]
Yes! 
![821BC9C3-E63C-4629-ACF2-757DEED571E0](https://user-images.githubusercontent.com/85020982/141115678-a0ab848c-5745-4da2-8437-fff25f1c3aa4.png)
![62019F18-649B-4FF1-943A-CF07A77C9E7C](https://user-images.githubusercontent.com/85020982/141115681-3fc41720-bff0-40d7-a700-97ae6af02e33.png)

**Comment by JakeScheck** [comment]
Indeed, my unit is also this version with the plasma pro feature, and auto mode.

**Comment by alockbox** [comment]
Also I have discord so that works too.

**Comment by RaresAil** [comment]
Alright, so remove the plugin from homebridge but keep its configuration and do the following:

1. On the machine you have the homebridge installed create a directory anywhere you want and inside that directory do the following:
```
// This will download the source files for the plugin in the same 
// folder where the command is executed, if it will create a new 
// folder just write "cd homebridge-levoit-air-purifier"
git clone https://github.com/RaresAil/homebridge-levoit-air-purifier.git .

// Switch to the branch that has the debug mode
git checkout feat/add-support-for-core-400s-pro

// Install the local modules for this plugin
npm install

// Build the plugin
npm build

// Link the plugin to homebridge
npm link
```

2. In the config for this accessory in homebridge add an extra parameter: `` "enableDebugMode": true``
```json
{
  "platforms": [
    {
      "name": "Levoit Air Purifiers",
      "email": "....",
      "password": "....",
      "platform": "LevoitAirPurifiers",
      "enableDebugMode": true
    }
  ]
}
```

3. Restart homebridge and look for the logs that contain the name of the plugin, e.g.
```
[10/11/2021, 17:13:38] [Levoit Air Purifiers] [DEBUG]: [LOGIN] The authentication success
[10/11/2021, 17:13:38] [Levoit Air Purifiers] Discovering devices...
[10/11/2021, 17:13:39] [Levoit Air Purifiers] [DEBUG]: [GET DEVICES] Device List -> JSON: ...
```

4. Send the output for the logs with ``[Levoit Air Purifiers]``

**Comment by RaresAil** [comment]
Yes, i will update the list to also accept 401S and because the 400S Pro model has 4 levels of speed (5 with the sleep mode) i will also make some changes to support all the levels of speed when the changes are done i will post here, also, to update the debug version, in the root of the plugin run:

```
git fetch
git pull
npm run build
npm link
```
And then restart homebridge, but that when i will add the new changes

**Comment by JakeScheck** [comment]
Fantastic! Thank you!

FYI, if you are interested in adding support for additional Levoit air purifiers, I also have a Levoit 131 which I just added back to Vesync. The debug log is below. This is a much less feature-rich unit, but would still be great to be able to turn on/off and adjust speed via Homebridge.

{"deviceRegion":"US","isOwner":true,"authKey":null,"deviceName":"Box","deviceImg":"https://image.vesync.com/defaultImages/LV_PUR131S_Series/icon_131_air_purifier_80.png","cid":"0MPRq4dGmjHyamgI8bDW2C1DEdiWmOsG","deviceStatus":"on","connectionStatus":"online","connectionType":"wifi","deviceType":"LV-PUR131S","type":"wifi-air","uuid":"319d8e2d-1a3e-4f3b-8ba2-ed085f73f2bd","configModule":"AirPurifier131","macID":null,"mode":"manual","speed":"low","extension":null,"currentFirmVersion":null,"subDeviceNo":null,"subDeviceType":null,"deviceFirstSetupTime":"Oct 3, 2021 7:42:59 PM"}

**Comment by RaresAil** [comment]
I am not sure if older version works with the API version that is used in this plugin, but after i will add the support for 400S Pro i will take a look

**Comment by JakeScheck** [comment]
sounds great! thanks again for your work! one of these days I'll dive into actual homebridge plugin development and not just use, but in the meantime, much appreciated!

**Comment by alockbox** [comment]
Wow you guys are fast. I'm at the office at the moment, and I run HOOBS actually.

**Comment by RaresAil** [comment]
@JakeScheck and @alockbox I pushed some changes that should detect the 400S, 400S Pro and 200S Air Purifiers and with the correct set for speed, please try the dev version and test all the features and tell me if everything is working 😃

**Comment by alockbox** [comment]
I was able to use the terminal in Hoobs to get all the steps done. I had to use SUDO for a few towards the end especially for the linking. I manually added the config back in after removing the public plug-in. I don't see anything in my logs after restarting the bridge. Something must be wrong specifically with Hoobs.

**Comment by RaresAil** [comment]
@alockbox You can remove the link by running `npm run remove-link` and it should work when the version will be available in homebridge, @JakeScheck should be able to test it and give us the feedback about it

**Comment by JakeScheck** [comment]
I can confirm, my 400S Pro is now functional! The control is a little strange but manageable, and maybe a quirk of the implementation of the API? My unit has 4 speed modes plus an auto mode and a sleep mode. When connected in HomeKit I have an on/off power button and a slider for speed control. The part that is odd to me is the auto mode and sleep mode are part of this slider. Essentially the slider has 6 “clicks” - auto, sleep, speed 1, 2, 3, 4. And when I slide the slider all the way down, it appears to also turn off. So the only way to get into auto mode is to slide the slider all the way down, then press the power button to turn it back on. This isn’t a huge issue to me as I normally just leave the thing on on speed 1 or 2 (and with automation capabilities I’ll primarily automate on/off). But if it was capable to implement sleep mode and auto mode as a linked switch or plug within HomeKit that would seem to be preferable for users that use those functions more. 

The air quality reading and PM2.5 value are working. 

Looking at the homebridge logs I did see some errors on initial connection. I will post that when I get to a computer this morning.

**Comment by RaresAil** [comment]
@JakeScheck Sleep mode should be the lowest speed like first one is speed 4, then is like: 4, 3, 2, 1, Sleep and 0 well is turn off. 

Yes the speed controll is wierd because HomeKit uses values from 0 to 100 and i mapped the 1, 2, 3, 4 and sleep mode to it, is like, Sleep mode is 20, 40, 60, 80, 100

And for Auto / Manual mode, you have to scroll even more where you will also have Control Lock and filter status
![IMG_4024](https://user-images.githubusercontent.com/47503490/141324684-acf29398-ce8b-4897-b876-935026cb29f1.PNG)
e:

**Comment by JakeScheck** [comment]
Thanks for the clarification on the fan speed control - makes sense (and I see now that if I had looked again at the documentation I would have understood this. Interesting, though, I don't see these additional elements in the Home app on my phone. I do see them in the Eve app, so it seems like the information is there, just not getting sent all the way through to Homekit?

Here are the logs from Homebridge. Perhaps the errors have something to do with the behavior I'm seeing. The notes below the "Adding new accessory..." item definitely seems related. The error that comes after that then repeats periodically in operation.
```
[11/11/2021, 10:13:55 AM] Loaded plugin: homebridge-levoit-air-purifier@1.1.0
[11/11/2021, 10:13:55 AM] Registering platform 'homebridge-levoit-air-purifier.LevoitAirPurifiers'
[11/11/2021, 10:13:55 AM] [Levoit Air Purifiers] Initializing LevoitAirPurifiers platform...
[11/11/2021, 10:13:55 AM] [Levoit Air Purifiers] [DEBUG]: [PLATFORM] Debug mode enabled
[11/11/2021, 10:13:55 AM] [Levoit Air Purifiers] Connecting to the servers...
[11/11/2021, 10:13:55 AM] [Levoit Air Purifiers] [DEBUG]: [START SESSION] Starting auth session...
[11/11/2021, 10:13:55 AM] [Levoit Air Purifiers] [DEBUG]: [LOGIN] Logging in...
[11/11/2021, 7:19:42 AM] [Levoit Air Purifiers] [DEBUG]: [LOGIN] The authentication success
[11/11/2021, 7:19:43 AM] [Levoit Air Purifiers] Discovering devices...
[11/11/2021, 7:19:43 AM] [Levoit Air Purifiers] [DEBUG]: [GET DEVICES] Device List -> JSON: [{"deviceRegion":"US","isOwner":true,"authKey":null,"deviceName":"Box","deviceImg":"https://image.vesync.com/defaultImages/LV_PUR131S_Series/icon_131_air_purifier_80.png","cid":"0MPRq4dGmjHyamgI8bDW2C1DEdiWmOsG","deviceStatus":"on","connectionStatus":"online","connectionType":"wifi","deviceType":"LV-PUR131S","type":"wifi-air","uuid":"319d8e2d-1a3e-4f3b-8ba2-ed085f73f2bd","configModule":"AirPurifier131","macID":null,"mode":"manual","speed":"low","extension":null,"currentFirmVersion":null,"subDeviceNo":null,"subDeviceType":null,"deviceFirstSetupTime":"Oct 3, 2021 7:42:59 PM"},{"deviceRegion":"US","isOwner":true,"authKey":null,"deviceName":"Tube","deviceImg":"https://image.vesync.com/defaultImages/Core_400S_Series/icon_core400s_purifier_160.png","cid":"vsaqc3b7aeb5489a89e592af43edcec2","deviceStatus":"on","connectionStatus":"online","connectionType":"WiFi+BTOnboarding+BTNotify","deviceType":"LAP-C401S-WUSR","type":"wifi-air","uuid":"48c9fafb-85c6-4a11-afbe-a1d982a3731b","configModule":"WiFiBTOnboardingNotify_AirPurifier_LAP-C401S-WUSR_US","macID":"78:e3:6d:9e:9c:d2","mode":null,"speed":null,"extension":{"airQuality":-1,"airQualityLevel":1,"mode":"manual","fanSpeedLevel":"1"},"currentFirmVersion":null,"subDeviceNo":null,"subDeviceType":null,"deviceFirstSetupTime":"Oct 3, 2021 7:30:52 PM"}]
[11/11/2021, 7:19:44 AM] [Levoit Air Purifiers] [DEBUG]: [GET DEVICE INFO] Getting device info...
[11/11/2021, 7:19:44 AM] [Levoit Air Purifiers] Adding new accessory: Tube
[Tube@@Filter Change Indication] Characteristic not in required or optional characteristic section for service AirPurifier. Adding anyway.
[Tube@@Filter Life Level] Characteristic not in required or optional characteristic section for service AirPurifier. Adding anyway.
[11/11/2021, 7:19:44 AM] Failed to save cached accessories to disk: Converting circular structure to JSON
    --> starting at object with constructor 'Platform'
    |     property 'registeredDevices' -> object with constructor 'Array'
    |     index 0 -> object with constructor 'VeSyncAccessory'
    --- property 'platform' closes the circle
[11/11/2021, 7:19:44 AM] Your accessories will not persist between restarts until this issue is resolved.
```

**Comment by RaresAil** [comment]
I will have to check that failed issue, is the first time i see that kind of issue 🤔

**Comment by RaresAil** [comment]
For home app, yeah is weird, there are not all the features every time

**Comment by RaresAil** [comment]
@JakeScheck i pushed an update that fixes that issue

**Comment by JakeScheck** [comment]
awesome! I'm kind of a git newbie - can you advise me on commands for updating the sideloaded plugin? This morning I just deleted the directory that I had done the git clone on - but there must be a more convenient way?

**Comment by RaresAil** [comment]
Well only with git clone, git fetch and pull, but if the previous version worked all good for you, i can release the new version that include the support for 400S

**Comment by RaresAil** [comment]
@JakeScheck @alockbox Is released as 1.1.0

**Comment by alockbox** [comment]
Thank you! They both now show up, amazing!

The only thing I can’t do is show them as separate tiles to make switching modes faster but it’s wonderful!

**Comment by JakeScheck** [comment]
Fantastic! I've updated to 1.1.0, and can confirm the errors that I saw previously have been taken care of.

I can also confirm that I can access the controls for Mode and Child Lock as well as the Filter information from within the Accessories > Air Purifier. And in fact when I have "Include in Home Status" set, and I hard-press the icon in the Status along the top, I get the full interface with all the controls and info. Great!

Now about my 131.... :) Kidding, unless you're interested. Thanks for all the efforts, this is awesome!
